File: src/domain/validators.ts
/**
 * Domain validators for hostie
 * 
 * Implements validation rules for hostnames, IPs, and other domain entities.
 */

import type { Entry } from "./types";

/**
 * Validation result type
 */
export interface ValidationResult {
  valid: boolean;
  error?: string;
}

/**
 * Validates a hostname according to RFC 952 and RFC 1123.
 * 
 * Rules:
 * - Total length: 1-255 characters
 * - Each label (segment between dots): 1-63 characters
 * - Characters: a-z, A-Z, 0-9, hyphen (-), period (.)
 * - First character of each label: letter or digit (RFC 1123 relaxation)
 * - Last character of each label: letter or digit (not hyphen or period)
 * - No consecutive periods
 * - Case insensitive
 * 
 * @param hostname - The hostname to validate
 * @returns ValidationResult with valid flag and optional error message
 */
export function validateHostname(hostname: string): ValidationResult {
  // Check for empty string
  if (!hostname || hostname.length === 0) {
    return {
      valid: false,
      error: "Hostname cannot be empty",
    };
  }

  // Check total length (1-255 characters)
  if (hostname.length > 255) {
    return {
      valid: false,
      error: "Hostname exceeds maximum length of 255 characters",
    };
  }

  // Check for consecutive periods
  if (hostname.includes("..")) {
    return {
      valid: false,
      error: "Hostname cannot contain consecutive periods",
    };
  }

  // Check for leading or trailing period
  if (hostname.startsWith(".")) {
    return {
      valid: false,
      error: "Hostname cannot start with a period",
    };
  }

  if (hostname.endsWith(".")) {
    return {
      valid: false,
      error: "Hostname cannot end with a period",
    };
  }

  // Split into labels and validate each
  const labels = hostname.split(".");

  for (const label of labels) {
    // Check label length (1-63 characters)
    if (label.length === 0) {
      return {
        valid: false,
        error: "Hostname labels cannot be empty",
      };
    }

    if (label.length > 63) {
      return {
        valid: false,
        error: `Hostname label "${label}" exceeds maximum length of 63 characters`,
      };
    }

    // Check first character: must be letter or digit (RFC 1123)
    const firstChar = label[0];
    if (!/[a-zA-Z0-9]/.test(firstChar)) {
      return {
        valid: false,
        error: `Hostname label "${label}" must start with a letter or digit`,
      };
    }

    // Check last character: must be letter or digit (not hyphen)
    const lastChar = label[label.length - 1];
    if (!/[a-zA-Z0-9]/.test(lastChar)) {
      return {
        valid: false,
        error: `Hostname label "${label}" must end with a letter or digit`,
      };
    }

    // Check all characters: must be letter, digit, or hyphen
    for (let i = 0; i < label.length; i++) {
      const char = label[i];
      if (!/[a-zA-Z0-9-]/.test(char)) {
        return {
          valid: false,
          error: `Hostname label "${label}" contains invalid character "${char}"`,
        };
      }
    }
  }

  // All checks passed
  return {
    valid: true,
  };
}

/**
 * Validates an IPv4 address in dotted-decimal notation.
 * 
 * Rules:
 * - Must have exactly 4 octets separated by periods
 * - Each octet must be a number between 0 and 255
 * - No leading zeros (except for "0" itself)
 * 
 * @param ip - The IPv4 address to validate
 * @returns ValidationResult with valid flag and optional error message
 */
export function validateIPv4(ip: string): ValidationResult {
  // Check for empty string
  if (!ip || ip.length === 0) {
    return {
      valid: false,
      error: "IPv4 address cannot be empty",
    };
  }

  // Split into octets
  const octets = ip.split(".");

  // Must have exactly 4 octets
  if (octets.length !== 4) {
    return {
      valid: false,
      error: "IPv4 address must have exactly four octets",
    };
  }

  // Validate each octet
  for (let i = 0; i < octets.length; i++) {
    const octet = octets[i];

    // Check if octet is numeric
    if (!/^\d+$/.test(octet)) {
      return {
        valid: false,
        error: `Octet "${octet}" must be numeric`,
      };
    }

    // Parse as integer
    const value = parseInt(octet, 10);

    // Check range (0-255)
    if (value < 0 || value > 255) {
      return {
        valid: false,
        error: `Octet "${octet}" must be between 0 and 255`,
      };
    }

    // Check for leading zeros (except "0" itself)
    if (octet.length > 1 && octet[0] === "0") {
      return {
        valid: false,
        error: `Octet "${octet}" cannot have leading zeros`,
      };
    }
  }

  // All checks passed
  return {
    valid: true,
  };
}

/**
 * Validates an IPv6 address in standard or compressed notation.
 * 
 * Rules:
 * - Up to 8 groups of 1-4 hexadecimal digits separated by colons
 * - Can use :: to compress consecutive zero groups (only once)
 * - Case insensitive
 * 
 * @param ip - The IPv6 address to validate
 * @returns ValidationResult with valid flag and optional error message
 */
export function validateIPv6(ip: string): ValidationResult {
  // Check for empty string
  if (!ip || ip.length === 0) {
    return {
      valid: false,
      error: "IPv6 address cannot be empty",
    };
  }

  // Check for triple or more consecutive colons
  if (ip.includes(":::")) {
    return {
      valid: false,
      error: "IPv6 address cannot have three or more consecutive colons",
    };
  }

  // Check for multiple :: compressions
  const doubleColonCount = (ip.match(/::/g) || []).length;
  if (doubleColonCount > 1) {
    return {
      valid: false,
      error: "IPv6 address can only have one double-colon compression",
    };
  }

  // Handle :: compression
  let groups: string[];
  if (ip.includes("::")) {
    // Split on :: and handle each side
    const parts = ip.split("::");
    const leftGroups = parts[0] ? parts[0].split(":") : [];
    const rightGroups = parts[1] ? parts[1].split(":") : [];
    
    // Calculate how many zero groups are compressed
    const totalGroups = leftGroups.length + rightGroups.length;
    if (totalGroups > 8) {
      return {
        valid: false,
        error: "IPv6 address cannot have more than 8 groups",
      };
    }
    
    // Combine for validation (we just need to validate the non-zero groups)
    groups = [...leftGroups, ...rightGroups].filter(g => g.length > 0);
  } else {
    // No compression, must have exactly 8 groups
    groups = ip.split(":");
    if (groups.length !== 8) {
      return {
        valid: false,
        error: "IPv6 address without compression must have exactly 8 groups",
      };
    }
  }

  // Validate each group
  for (const group of groups) {
    // Each group must be 1-4 hexadecimal digits
    if (group.length === 0 || group.length > 4) {
      return {
        valid: false,
        error: `IPv6 group "${group}" must be 1-4 hexadecimal digits`,
      };
    }

    // Check if all characters are valid hex
    if (!/^[0-9a-fA-F]+$/.test(group)) {
      return {
        valid: false,
        error: `IPv6 group "${group}" must contain only hexadecimal digits`,
      };
    }
  }

  // All checks passed
  return {
    valid: true,
  };
}

/**
 * Validates an IP address (either IPv4 or IPv6).
 * 
 * @param ip - The IP address to validate
 * @returns ValidationResult with valid flag and optional error message
 */
export function validateIP(ip: string): ValidationResult {
  // Try IPv4 first
  const ipv4Result = validateIPv4(ip);
  if (ipv4Result.valid) {
    return ipv4Result;
  }

  // Try IPv6
  const ipv6Result = validateIPv6(ip);
  if (ipv6Result.valid) {
    return ipv6Result;
  }

  // Neither worked, return a generic error
  return {
    valid: false,
    error: "Invalid IP address (neither valid IPv4 nor IPv6)",
  };
}

/**
 * Validates that no two enabled entries have duplicate hostnames or aliases.
 * 
 * Rules:
 * - Only checks enabled entries (disabled entries can have duplicates)
 * - Case-insensitive comparison
 * - Checks both hostnames and aliases
 * - A hostname cannot conflict with another hostname or alias
 * 
 * @param entries - Array of entries to validate
 * @returns ValidationResult with valid flag and optional error message
 */
export function validateNoDuplicates(entries: Entry[]): ValidationResult {
  // Filter to only enabled entries
  const enabledEntries = entries.filter(entry => entry.enabled);

  // If 0 or 1 enabled entries, no duplicates possible
  if (enabledEntries.length <= 1) {
    return {
      valid: true,
    };
  }

  // Collect all hostnames and aliases (case-insensitive)
  const seenNames = new Map<string, string>(); // lowercase -> original

  for (const entry of enabledEntries) {
    // Check hostname
    const hostnameLower = entry.hostname.toLowerCase();
    if (seenNames.has(hostnameLower)) {
      return {
        valid: false,
        error: `Duplicate hostname "${hostnameLower}" found in enabled entries`,
      };
    }
    seenNames.set(hostnameLower, entry.hostname);

    // Check aliases
    for (const alias of entry.aliases) {
      const aliasLower = alias.toLowerCase();
      if (seenNames.has(aliasLower)) {
        return {
          valid: false,
          error: `Duplicate hostname/alias "${aliasLower}" found in enabled entries`,
        };
      }
      seenNames.set(aliasLower, alias);
    }
  }

  // All checks passed
  return {
    valid: true,
  };
}

File: go/internal/domain/validators.go
// Validators for hostie domain types.
//
// Ports v1 src/domain/validators.ts to Go. Returns error (nil = valid).
package domain

import (
	"fmt"
	"net"
	"strconv"
	"strings"
)

// ValidateHostname checks RFC 952/1123 compliance.
//
// Rules (mirrors v1):
//   - 1..255 total chars
//   - no leading/trailing dot, no consecutive dots
//   - labels 1..63 chars
//   - each label: first/last char alnum, body alnum or hyphen
func ValidateHostname(hostname string) error {
	if hostname == "" {
		return fmt.Errorf("hostname cannot be empty")
	}
	if len(hostname) > 255 {
		return fmt.Errorf("hostname exceeds maximum length of 255 characters")
	}
	if strings.Contains(hostname, "..") {
		return fmt.Errorf("hostname cannot contain consecutive periods")
	}
	if strings.HasPrefix(hostname, ".") {
		return fmt.Errorf("hostname cannot start with a period")
	}
	if strings.HasSuffix(hostname, ".") {
		return fmt.Errorf("hostname cannot end with a period")
	}

	for _, label := range strings.Split(hostname, ".") {
		if label == "" {
			return fmt.Errorf("hostname labels cannot be empty")
		}
		if len(label) > 63 {
			return fmt.Errorf("hostname label %q exceeds maximum length of 63 characters", label)
		}
		if !isAlnum(label[0]) {
			return fmt.Errorf("hostname label %q must start with a letter or digit", label)
		}
		if !isAlnum(label[len(label)-1]) {
			return fmt.Errorf("hostname label %q must end with a letter or digit", label)
		}
		for i := 0; i < len(label); i++ {
			c := label[i]
			if !isAlnum(c) && c != '-' {
				return fmt.Errorf("hostname label %q contains invalid character %q", label, string(c))
			}
		}
	}
	return nil
}

func isAlnum(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')
}

// ValidateIPv4 checks dotted-decimal: 4 octets 0..255, no leading zeros, no whitespace.
func ValidateIPv4(ip string) error {
	if ip == "" {
		return fmt.Errorf("IPv4 address cannot be empty")
	}
	if ip != strings.TrimSpace(ip) {
		return fmt.Errorf("IPv4 address must not have leading or trailing whitespace")
	}
	octets := strings.Split(ip, ".")
	if len(octets) != 4 {
		return fmt.Errorf("IPv4 address must have exactly four octets")
	}
	for _, oct := range octets {
		if oct == "" {
			return fmt.Errorf("octet %q must be numeric", oct)
		}
		for i := 0; i < len(oct); i++ {
			if oct[i] < '0' || oct[i] > '9' {
				return fmt.Errorf("octet %q must be numeric", oct)
			}
		}
		v, err := strconv.Atoi(oct)
		if err != nil {
			return fmt.Errorf("octet %q must be numeric", oct)
		}
		if v < 0 || v > 255 {
			return fmt.Errorf("Octet %q must be between 0 and 255", oct)
		}
		if len(oct) > 1 && oct[0] == '0' {
			return fmt.Errorf("octet %q cannot have leading zeros", oct)
		}
	}
	return nil
}

// ValidateIPv6 checks IPv6 syntax (mirrors v1: rejects IPv4-mapped & dotted-quad embeds).
//
// We hand-roll the parser to match v1 semantics exactly rather than rely on
// net.ParseIP, which accepts IPv4-mapped forms ("::ffff:1.2.3.4") that v1 rejects.
func ValidateIPv6(ip string) error {
	if ip == "" {
		return fmt.Errorf("IPv6 address cannot be empty")
	}
	if strings.Contains(ip, ":::") {
		return fmt.Errorf("IPv6 address cannot have three or more consecutive colons")
	}
	if strings.Count(ip, "::") > 1 {
		return fmt.Errorf("IPv6 address can only have one double-colon compression")
	}

	var groups []string
	if strings.Contains(ip, "::") {
		parts := strings.SplitN(ip, "::", 2)
		var left, right []string
		if parts[0] != "" {
			left = strings.Split(parts[0], ":")
		}
		if parts[1] != "" {
			right = strings.Split(parts[1], ":")
		}
		// With `::` present, the elided run must cover ≥1 zero group,
		// so total explicit groups must be < 8.
		if len(left)+len(right) >= 8 {
			return fmt.Errorf("IPv6 address cannot have more than 8 groups")
		}
		groups = append(groups, left...)
		groups = append(groups, right...)
	} else {
		groups = strings.Split(ip, ":")
		if len(groups) != 8 {
			return fmt.Errorf("IPv6 address without compression must have exactly 8 groups")
		}
	}

	for _, g := range groups {
		if len(g) == 0 || len(g) > 4 {
			return fmt.Errorf("IPv6 group %q must be 1-4 hexadecimal digits", g)
		}
		for i := 0; i < len(g); i++ {
			c := g[i]
			if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')) {
				return fmt.Errorf("IPv6 group %q must contain only hexadecimal digits", g)
			}
		}
	}

	// Sanity: net.ParseIP must also accept it (defensive parity).
	if parsed := net.ParseIP(ip); parsed == nil {
		return fmt.Errorf("invalid IPv6 address %q", ip)
	}
	return nil
}

// ValidateIP accepts either valid IPv4 or IPv6 (v1 parity).
func ValidateIP(ip string) error {
	if err := ValidateIPv4(ip); err == nil {
		return nil
	}
	if err := ValidateIPv6(ip); err == nil {
		return nil
	}
	return fmt.Errorf("invalid IP address (neither valid IPv4 nor IPv6)")
}

// ValidateComment checks that a comment does not contain newlines or control
// characters that could break out of the managed block or inject malicious content.
func ValidateComment(comment string) error {
	for i := 0; i < len(comment); i++ {
		c := comment[i]
		// Reject newline (\n, \r) and all ASCII control characters (0x00-0x1F, 0x7F)
		if c == '\n' || c == '\r' || c < 0x20 || c == 0x7F {
			return fmt.Errorf("comment contains invalid control character at position %d", i)
		}
	}
	return nil
}

// ValidateNoDuplicates returns the first duplicate hostname/alias collision
// across enabled entries. Case-insensitive (v1 parity). Disabled entries are
// ignored entirely.
func ValidateNoDuplicates(entries []Entry) error {
	seen := make(map[string]struct{})
	for _, e := range entries {
		if !e.Enabled {
			continue
		}
		h := strings.ToLower(e.Hostname)
		if _, dup := seen[h]; dup {
			return fmt.Errorf("duplicate hostname %q found in enabled entries", h)
		}
		seen[h] = struct{}{}
		for _, a := range e.Aliases {
			al := strings.ToLower(a)
			if _, dup := seen[al]; dup {
				return fmt.Errorf("duplicate hostname/alias %q found in enabled entries", al)
			}
			seen[al] = struct{}{}
		}
	}
	return nil
}
